File: Java/Irrgarten/src/irrgarten/Game.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

/**
 * Class representing the labyrinth game.
 */
public class Game {
    private static int MAX_ROUNDS = 10;
    private int currentPlayerIndex;
    private String log;
    private Player currentPlayer;
    private Labyrinth labyrinth;
    private ArrayList<Player> players = new ArrayList<>();
    private ArrayList<Monster> monsters = new ArrayList<>();
    
    /**
     * Constructor for the Game class.
     * @param nPlayers Number of players in the game.
     */
    public Game(int nPlayers){
        // Labyrinth setup
        int nRows = 9;
        int nCols = 9;
        int exitRow = 7;
        int exitCol = 8;
        labyrinth = new Labyrinth(nRows,nCols,exitRow,exitCol);
        
        // Creating players
        for (int i = 0; i < nPlayers; i++) {
            Player player = new Player((char)('0' + i ), Dice.randomIntelligence(), Dice.randomStrenght());
            players.add(player);
        }
        Player player = new Player((char)('0' + 1 ), 0, 0);
        players.add(player);
        // Creating monsters
        for (int i = 0; i < nPlayers; i++) {
            Monster monster = new Monster("Monster " + i , Dice.randomIntelligence(), Dice.randomStrenght());
            monsters.add(monster);
        }
        Monster monster = new Monster("Boss " , 10, 10);
        monsters.add(monster);
        // Determine who starts
        currentPlayerIndex = Dice.whoStarts(nPlayers);
        currentPlayer = players.get(currentPlayerIndex);
        
        // Initialize other attributes
        log = "";

        // Configure the labyrinth
        configureLabyrinth();
    }
    
    /**
     * Checks if the game has finished.
     * @return true if the game has finished, false otherwise.
     */
    public boolean finished(){
        return labyrinth.haveAWinner();
    }
    
    /**
     * Performs the next step in the game.
     * @param preferredDirection Preferred direction to move.
     * @return true if the game has finished after the step, false otherwise.
     */
    public boolean nextStep(Directions preferredDirection){
        log = "";
        boolean dead = currentPlayer.dead();
        if(!dead){
            Directions direction = actualDirection(preferredDirection);
            if(direction != preferredDirection){
                logPlayerNoOrders();
            }
            Monster monster = labyrinth.putPlayer(direction,currentPlayer);
            
            if(monster == null){
                logNoMonster();
            }
            else{
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }
        }
        else{
            manageResurrection();
        }
        
        boolean endGame = finished();
        
        if(!endGame){
            nextPlayer();
        }
        
        return endGame;
    }
    
    /**
     * Gets the current state of the game.
     * @return Current state of the game.
     */
    public GameState getGameState(){
        String namesplayers = "";
        String namesmonsters = "";
        for(Player player : players){
            namesplayers += player.toString() + "\n";
        }
        for(Monster monster : monsters){
            namesmonsters += monster.toString() + "\n";
        }
        return new GameState(labyrinth.toString(), namesplayers, namesmonsters, currentPlayerIndex, this.finished(),log);
    }
    
    /**
     * Configures the labyrinth with blocks and monsters.
     */
    private void configureLabyrinth() {
        // Add blocks
        labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, 9);
        labyrinth.addBlock(Orientation.HORIZONTAL, 8, 0, 9);
        labyrinth.addBlock(Orientation.HORIZONTAL, 5, 1, 2);
        labyrinth.addBlock(Orientation.HORIZONTAL, 6, 2, 3);
        labyrinth.addBlock(Orientation.HORIZONTAL, 2, 4, 2);
        labyrinth.addBlock(Orientation.HORIZONTAL, 4, 4, 2);
        labyrinth.addBlock(Orientation.VERTICAL, 1, 0, 5);
        labyrinth.addBlock(Orientation.VERTICAL, 2, 2, 3);
        labyrinth.addBlock(Orientation.VERTICAL, 1, 8, 6);
        labyrinth.addBlock(Orientation.VERTICAL, 2, 6, 7);
        
        // Add monsters
        labyrinth.addMonster(4, 1, monsters.get(0));
        labyrinth.addMonster(7, 7, monsters.get(1));
        //labyrinth.addMonster(3, 4, monsters.get(2));
        //labyrinth.addMonster(5, 4, monsters.get(3));
         
        // Spread players in the labyrinth
        labyrinth.spreadPlayers(players);
    }
    
    /**
     * Moves to the next player in turn.
     */
    public void nextPlayer(){
        int index = (currentPlayerIndex + 1) % players.size();
        currentPlayerIndex = index;
        currentPlayer = players.get(currentPlayerIndex);
    }
    
    /**
     * Gets the current direction for the player.
     * @param preferredDirection Preferred direction to move.
     * @return Current direction for the player.
     */
    private Directions actualDirection(Directions preferredDirection){
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();
        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, currentCol);
        return currentPlayer.move(preferredDirection, validMoves);
    }
    
    /**
     * Simulates a combat between the player and a monster.
     * @param monster Monster to combat against.
     * @return The winner of the combat.
     */
    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        while(!lose && (rounds < MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds++;
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            if(!lose){
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        logRounds(rounds,MAX_ROUNDS);
        return winner;
    }
    
    /**
     * Manages the reward after a combat.
     * @param winner The winner of the combat.
     */
    private void manageReward(GameCharacter winner){
        if(winner == GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            logPlayerWon();
        }
        else{
            logMonsterWon();
        }
    }
    
    /**
     * Manages player resurrection.
     */
    private void manageResurrection(){
        boolean resurrect = Dice.resurrectPlayer();
        if(resurrect){
            currentPlayer.resurrect();
            FuzzyPlayer fuzzyPlayer = new FuzzyPlayer(currentPlayer);
            players.set(currentPlayerIndex, fuzzyPlayer);
            labyrinth.setFuzzyPlayer(currentPlayer.getRow(), currentPlayer.getCol(), fuzzyPlayer);
            currentPlayer = fuzzyPlayer;
            logResurrected();
        }
        else{
            logPlayerSkipTurn();
        }   
    }
    
    // Methods to log events in the game log
    
    private void logPlayerWon(){
        log += "Player " + currentPlayerIndex + " wins the combat" + "\n" + currentPlayer;
    }
    
    private void logMonsterWon(){
        log += "Monster wins the combat" + "\n";
    }
    
    private void logResurrected(){
        log +=  "Player " + currentPlayerIndex +  " resurrects" + "\n";
    }
    
    private void logPlayerSkipTurn(){
        log += "Player " + currentPlayerIndex + " lost the turn" + "\n";
    }
    
    private void logPlayerNoOrders(){
        log += "Player " + currentPlayerIndex + " couldnt follow the orders" + "\n";
    }
    
    private void logNoMonster(){
        log += "Player " + currentPlayerIndex + " get into an empty square" + "\n";
    }
    
    private void logRounds(int rounds, int max){
        log += "Reached round " + rounds + ", Max rounds  " + max + "\n";
    }
    
}

File: Ruby/Game.rb
#encoding utf-8

require_relative 'GameState'



module Irrgarten
    class Game
        @@MAX_ROUNDS = 10
        
        def initialize(n_players)
            @labyrinth = Labyrinth.new(5,5,4,4)
            @log="Game starts"
            @players=Array.new(n_players)
            @monsters=Array.new(n_players)

            n_players.times do |i|
                @players << Player.new(i, Dice.random_intelligence, Dice.random_strenght)
            end


            n_players.times do |i|
                @monsters << Monster.new("Monster" + i.to_s, Dice.random_intelligence, Dice.random_strenght)
            end

            @current_player_index=Dice.who_starts(n_players)
            @current_player=@players[@current_player_index]

            configure_labyrinth()

        end

        def finished
            return @labyrinth.have_a_winner
        end

        def next_step(preferredDirection)
            dead = @current_player.dead
            if(!dead)
                direction = actual_direction(preferredDirection)
                if(direction != preferredDirection)
                    log_player_no_orders()
                end
                monster = @labyrinth.put_player(direction, @current_player)
                if(monster == nil)
                    log_no_monster()
                else
                    winner = combat(monster)
                    manage_reward(winner)
                end
            else
                manage_resurrection()
            end
            end_game = finished()
            if(!end_game)
                next_player()
            end
        end

        def get_game_state
            string_players = ""
            string_monsters = ""
            @players.each do |player|
                string_players += player.to_s
            end
            @monsters.each do |monster|
                string_monsters += monster.to_s
            end

            return GameState.new(@labyrinth.to_s, string_players, string_monsters, @current_player, finished(), @log)
        end

        def configure_labyrinth
            n_rows=5
            n_cols=5
            manual_labyrinth = [
                ['X', 'X', 'X', 'X', 'X'],
                ['X', '-', '-', '-', 'X'],
                ['X', '-', 'X', '-', 'X'],
                ['X', '-', '-', 'E', 'X'],
                ['X', 'X', 'X', 'X', 'X']
              ]
            n_rows.times do |i|
                n_cols.times do |j|
                    @labyrinth.set_square(i, j, manual_labyrinth[i][j])
                end
            end
        end

        def next_player
            @current_player_index += 1
            @current_player=players[@current_player_index]
        end

        def actual_direction(preferredDirection)
            current_row = @current_player.row
            current_col = @current_player.col
            valid_moves = @labyrinth.valid_moves(current_row, current_col)
            output = @current_player.move(preferredDirection, valid_moves)
            return output
        end

        def combat(monster)
            rounds = 0
            winner = GameCharacter::PLAYER
            player_attack = @current_player.attack
            lose = monster.defend(player_attack)
            while(!lose && rounds < @@MAX_ROUNDS)
                rounds += 1
                winner = GameCharacter::MONSTER
                monster_attack = monster.attack
                lose = @current_player.defend(monster_attack)
                if(!lose)
                    player_attack = @current_player.attack
                    winner = GameCharacter::PLAYER
                    lose = monster.defend(player_attack)
                end
                log_rounds(rounds, @@MAX_ROUNDS)
            end
            return winner
        end

        def manage_reward(winner)
            if(winner == GameCharacter::PLAYER)
                receive_reward()
                log_player_won()
            else
                log_monster_won()
            end

        end

        def manage_resurrection
            resurrect = Dice.resurrect_player
            if(resurrect)
                @current_player.resurrect
                log_resurrected()
            else
                log_player_skip_turn()
            end
        end

        def log_player_won
            @log += "Player wins the combat" + "\n";
        end

        def log_monster_won
            @log += "Monster wins the combat" + "\n";
        end

        def log_resurrected
            @log += "Player resurrects" + "\n";
        end

        def log_player_skip_turn
            @log += "Player lost the turn" + "\n";
        end

        def log_player_no_orders
            @log += "Player couldnt follow the orders" + "\n";
        end

        def log_no_monster
            @log += "Player get into an empty square" + "\n";
        end

        def log_rounds(rounds, max)
            @log += "Max rounds achieved" + "\n";
        end

    end
end